dot/types/digest.go

```diff
@@ -5,24 +5,123 @@ package types
 
 import (
 	"fmt"
+	"strings"
 
 	"github.com/ChainSafe/gossamer/pkg/scale"
 )
 
 // DigestItem is a varying date type that holds type identifier and a scaled encoded message payload.
-type DigestItem struct {
+type digestItem struct {
 	ConsensusEngineID ConsensusEngineID
 	Data              []byte
 }
 
+type DigestItem struct {
+	inner any
+}
+
+type DigestItemValues interface {
+	PreRuntimeDigest | ConsensusDigest | SealDigest | RuntimeEnvironmentUpdated
+}
+
+func newDigestItem[Value DigestItemValues](value Value) DigestItem {
+	item := DigestItem{}
+	setDigestItem[Value](&item, value)
+	return item
+}
+
+func setDigestItem[Value DigestItemValues](mvdt *DigestItem, value Value) {
+	mvdt.inner = value
+}
+
+func (mvdt *DigestItem) SetValue(value any) (err error) {
+	switch value := value.(type) {
+	case PreRuntimeDigest:
+		setDigestItem(mvdt, value)
+		return
+	case ConsensusDigest:
+		setDigestItem(mvdt, value)
+		return
+	case SealDigest:
+		setDigestItem(mvdt, value)
+		return
+	case RuntimeEnvironmentUpdated:
+		setDigestItem(mvdt, value)
+		return
+	default:
+		return fmt.Errorf("unsupported type")
+	}
+}
+
+func (mvdt DigestItem) IndexValue() (index uint, value any, err error) {
+	switch mvdt.inner.(type) {
+	case PreRuntimeDigest:
+		return 6, mvdt.inner, nil
+	case ConsensusDigest:
+		return 4, mvdt.inner, nil
+	case SealDigest:
+		return 5, mvdt.inner, nil
+	case RuntimeEnvironmentUpdated:
+		return 8, mvdt.inner, nil
+	}
+	return 0, nil, scale.ErrUnsupportedVaryingDataTypeValue
+}
+
+func (mvdt DigestItem) Value() (value any, err error) {
+	_, value, err = mvdt.IndexValue()
+	return
+}
+
+func (mvdt DigestItem) ValueAt(index uint) (value any, err error) {
+	switch index {
+	case 6:
```

Comment:
Nit/should we list these in order?

Reply:
I just kept this based on the order the types were put in the file.